### DISCUSSION TITLE
Guest reactions

### DISCUSSION THREAD

**Discussion by MWStudio4** [cybercog/laravel-love#23] [Q&A] [answered]
Hello. As i understand there is no guest likes. At the moment i provide user_id=1 . It would be nice to get user_id nullable. But i don't know if it affects other methods

**Comment by antonkomarev** [comment]
@MWStudio4 you are right. Current implementation isn't designed for guest likes out of the box.

Passing `null` to `user_id` wouldn't solve this task, because it will affect on any other guest users.

I see 3 possible solutions here:
1. When guest likes any models you could save them on the client side or in session, and after successful registration or authentication you can apply this likes one by one.
2. Any guest session should have model `Guest` with unique identifier, and then `Like` model should be linked with polymorpic `liker` instead of `user_id`. And after registration or authentication you can fetch all Guest model likes and change `liker` to required User model.
3. Any guest session should have model `Guest` with unique identifier, and then in `Like` model you can make `user_id` nullable and add nullable `guest_id`. If guest will like model - you will fill `guest_id` field, and if authenticated user likes model you will fill `user_id` field. And after registration or authentication you can fetch all Guest model likes and set `user_id` to required User model id and set `null` to `guest_id` fields.

Just don't forget that if you allows guest to like models there could be issues with counters cheating.
I'd prefer first solution because it wouldn't affect on counters and will be visible for the guest only until he will be authenticated.

**Comment by MWStudio4** [comment]
I see thanks. And surely about cheating. To say the truth I doubt about need of guest likes at the moment. About cheating may be there will be good way to implement rate limit

**Comment by antonkomarev** [comment]
Some more thoughts for guest reactions:

- Guest Reacter can react to entity and this value should be temporary saved but displayed only for this Reacter.
- If Guest authenticates and we cannot find such reaction from authenticated user in database - we should persist it, otherwise we should delete temporary reaction.

**Comment by MWStudio4** [comment]
Nice thoughts. Thanks!

**Comment by antonkomarev** [comment, marked as answer]
Since v6 any model could act as `Reacter`. That means you can create `Guest` Eloquent model and make it `Reacterable`.

After authentication you can collect all guest's reactions and change their `reacter_id` on authenticated user's reacter id. Just don't forget to check if model was already reacted by user's reacter. Counters of affected reactants should be rebuilt.

**Comment by MWStudio4** [comment]
Good news! Thank you.

**Comment by vlados** [comment]
Can someone post example implementation of this?

**Comment by antonkomarev** [comment]
@vlados it depends from your system requirements and application logic a lot. But simple implementation might look like this one.

Add `GuestUser` model:
```php
<?php

namespace App;

use Cog\Laravel\Love\Reacterable\Models\Traits\Reacterable;
use Cog\Contracts\Love\Reacterable\Models\Reacterable as ReacterableInterface;

class GuestUser extends Model implements
    ReacterableInterface
{
    use Reacterable;

    protected $table = 'guest_users';
}
```

When user is visiting your page - you are creating `GuestUser` model instance and bind it to the session:
```php
$guestUser = GuestUser::create();
```

Then guest user adding some likes:
```php
$guestUser->viaLoveReacter()->reactTo($goodArticle, 'like');
$guestUser->viaLoveReacter()->reactTo($badArticle, 'dislike');
```

On successful login you are going though the collection of guest user reactions and adding them to user and then deleting this GuestUser or only his reactions. For this you might store latest GuestUser in User as attribute.

```php
$user = $request->user();
$guestUserReacter = $user->getGuestUser()->viaLoveReacter();
$userReacter = $user->viaLoveReacter();

$reactions = $guestUserReacter->getReactions();
foreach ($reactions as $reaction) {
    $reactable = $reaction->getReactant()->getReactable();
    $reactionTypeName = $reaction->getReactionType()->getName();
    $reactionRate = $reaction->getRate();

    $userReacter->reactTo($reactable, $reactionTypeName, $reactionRate);

    $guestUserReacter->unreactTo($reactable, $reactionTypeName);
}
```
